rust2/core/src/error.rs

```diff
@@ -0,0 +1,166 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Error, status and result types.
+
+use std::{ffi::NulError, fmt::Display};
+
+use arrow::error::ArrowError;
+
+/// Status of an operation.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Status {
+    /// No error.
+    Ok,
+    /// An unknown error occurred.
+    Unknown,
+    /// The operation is not implemented or supported.
+    NotImplemented,
+    /// A requested resource was not found.
+    NotFound,
+    /// A requested resource already exists.
+    AlreadyExists,
+    /// The arguments are invalid, likely a programming error.
+    /// For instance, they may be of the wrong format, or out of range.
+    InvalidArguments,
+    /// The preconditions for the operation are not met, likely a programming error.
+    /// For instance, the object may be uninitialized, or may have not
+    /// been fully configured.
+    InvalidState,
+    /// Invalid data was processed (not a programming error).
+    /// For instance, a division by zero may have occurred during query
+    /// execution.
+    InvalidData,
+    /// The database's integrity was affected.
+    /// For instance, a foreign key check may have failed, or a uniqueness
+    /// constraint may have been violated.
+    Integrity,
+    /// An error internal to the driver or database occurred.
+    Internal,
+    /// An I/O error occurred.
+    /// For instance, a remote service may be unavailable.
+    IO,
+    /// The operation was cancelled, not due to a timeout.
+    Cancelled,
+    /// The operation was cancelled due to a timeout.
+    Timeout,
+    /// Authentication failed.
+    Unauthenticated,
+    /// The client is not authorized to perform the given operation.
+    Unauthorized,
+}
+
+/// An ADBC error.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Error {
+    /// The error message.
+    pub message: String,
+    /// The status of the operation.
+    pub status: Status,
+    /// A vendor-specific error code, if applicable.
+    pub vendor_code: i32,
+    /// A SQLSTATE error code, if provided, as defined by the SQL:2003 standard.
+    /// If not set, it should be set to `\0\0\0\0\0`.
+    pub sqlstate: [i8; 5],
```

Comment:
Because `c_char` [translates](https://doc.rust-lang.org/std/ffi/type.c_char.html) to `i8` on my platform. Maybe we can switch to `c_char`?

Reply:
Ah, I didn't consider the ffi use-case here. Maybe we can use `u8` here and convert to `c_char` only for ffi? (Because to support ffi directly here I think `message` should be `CString` I think? - and what about `details`?)

Reply:
In the full version there is code for converting between the two yes. You can see it here: 
- https://github.com/alexandreyc/adbc-rs/blob/main/core/src/ffi/types.rs#L488
- https://github.com/alexandreyc/adbc-rs/blob/main/core/src/ffi/types.rs#L545

Reply:
I switched to `c_char` for the moment and added a TODO to revisit it later. Currently `sqlstate` is only used when printing the error, so I guess it's not a big deal.